### DISCUSSION TITLE
Horiztonal text contatenation and proper way to render to ANSI.

### DISCUSSION THREAD

**Discussion by Erotemic** [Textualize/rich#3076] [Q&A] [answered]
I'm currently writing a PR for line-profiler to integrate rich as an optional dependency to make the output prettier. So far it is looking good: 
https://github.com/pyutils/line_profiler/pull/225 However, there is a small corner case and I'm not sure if I'm doing things completely correctly.

To this end, I have a MWE that illustrates my questions.

I want to be able to:

* take two blocks of text with the same number of lines
* highlight the left-hand-side and right-hand-side independently
* horizontally concatentate the text
* generate the corresponding ANSI so I can write to a stream (e.g. stdout) directly. 

I've been able to do this using the following approach:

* Generate the highlighted block of text for the LHS and RHS.
* Use rich.columns.Columns to achieve horiztonal concatenation (is there a better way?) 
* Create a Console object to get the ANSI text I need to write to the stream. (is there a better way?)

To illustrate this I have an example where the RHS is a small block of code, and the LHS are line numbers.

```python
#https://github.com/pyutils/line_profiler/pull/225
import sys
from rich.syntax import Syntax
from rich.text import Text
from rich.columns import Columns
from rich.console import Console
from rich.highlighter import ReprHighlighter
import textwrap
import io

# The interface must write to a stream (e.g. sys.stdout)
stream = sys.stdout

# Example code for the right-hand-side
rhs_text = textwrap.dedent(
    '''
    def fib(n):
        a, b = 0, 1
        while a < n:
            print(a, end=' ')
            a, b = b, a+b
        print()
    fib(1000)
    ''').strip()

# Line numbers for the left-hand-side text
nlines = rhs_text.count('\n') + 1
lhs_text = '\n'.join(str(i).rjust(4) + ' | ' for i in range(1, nlines + 1))
```

My current code for processing these into the final output is:

```python
# Highlight and format the text
rhs = Syntax(rhs_text, 'python', background_color='default')
lhs = Text(lhs_text)
ReprHighlighter().highlight(lhs)
renderable = Columns([lhs, '', rhs])
tmp = io.StringIO()
write_console = Console(file=tmp, soft_wrap=True, color_system='standard')
write_console.print(renderable)
block = tmp.getvalue()

stream.write(block)
stream.write('\n')
```

This works when the terminal is sufficiently wide. I get a nice render:

![image](https://github.com/Textualize/rich/assets/3186211/82e51496-5205-4fa1-886f-819c497c692d)

However, if I make my terminal too small, the columns split the LHS and RHS onto separte lines:

![image](https://github.com/Textualize/rich/assets/3186211/daa1164a-827b-4801-8baf-90d87c70b20b)

Question 1: Is there a way to more directly horizontally concatenate highlighted text objects such that the final output strongly ties the LHS line to the RHS line?

Question 2: The way I'm extracting the ANSI representation of the highlighted text seems roundabout. Is this the standard way to do this, or is there a better way?

**Comment by willmcgugan** [comment, marked as answer]
Hi Jon.

The `Columns` widget is designed to work this way. If you don't want the wrapping behaviour you could use `Table.grid()` which creates a borderless table.

Have a look at [Capturing output](https://rich.readthedocs.io/en/latest/console.html#capturing-output) for an API to capture ansi text.

**Comment by Erotemic** [comment]
Perfect! Works like a charm.

Modified second block:

```python
from rich.table import Table
# Highlight and format the text
rhs = Syntax(rhs_text, 'python', background_color='default')
lhs = Text(lhs_text)
ReprHighlighter().highlight(lhs)
table = Table.grid()
table.add_row(lhs, ' ', rhs)
write_console = Console(file=stream, soft_wrap=True, color_system='standard')
write_console.print(table)
stream.write('\n')
```

Result:

![image](https://github.com/Textualize/rich/assets/3186211/8142cbed-d340-45d5-8e9c-ba72373015ae)
